src/board.rs

```diff
@@ -106,123 +215,175 @@ impl Chessboard {
-// }
+            // Add the row string to the FEN components vector
+            string_array.push(row_string);
+            string_array.push("/".to_owned());
+        }
 
-// en_passant_square() {
-//     // Return the square where an en passant capture is possible, or -1 if not possible
-//     // For example, if the last move was a double-step pawn move, return the square where the opponent's pawn can capture en passant.
-// }
+        let binding = string_array.concat();
+        fen_string = binding.chars();
+        fen_string.next_back();
+        let fen_string_no_slash = fen_string.as_str();
+        for item in &mut string_array {
+            item.clear();
+        }
+        string_array.push(fen_string_no_slash.to_owned());
 
-// castling_rights() {
-//     // Use flags or bitmasks to represent castling rights for both players
-//     // For example, if kingside castling is allowed for white, set white_kingside_castle = true
-//     // Similarly, maintain flags for black's castling rights
-// }
+        string_array.push(" ".to_owned());
 
-// make_move(move) {
-//     // Update the bitboards and game state based on the move
-//     // For example, if it's a pawn move, update the pawn bitboard
-//     // If it's a capture, clear the bit for the captured piece
-//     // Update en passant square and castling rights accordingly
-//     // Switch turns: white_turn = !white_turn;
-// }
+        // Whose turn
+        string_array.push(if self.white_turn {
+            "w ".to_string()
+        } else {
+            "b ".to_string()
+        });
 
-// generate_legal_moves() {
-//     legal_moves = []
-
-//     for each piece in pieces_of_current_turn {
-//         possible_moves = generate_moves_for_piece(piece)
-//         for each move in possible_moves {
-//             if is_legal_move(move) {
-//                 legal_moves.push(move)
-//             }
-//         }
-//     }
-//
-//     return legal_moves
-// }
+        // Castling rights
+        string_array.push(match self.castling_rights {
+            0 => "- ".to_string(),
+            rights => {
+                let mut rights_string = String::new();
 
-// generate_moves_for_piece(piece) {
-//     // Generate all possible moves for the given piece
-//     // Consider piece-specific movement rules (e.g., pawn's initial double move, castling for king, etc.)
-// }
+                // Check individual castling rights and append to rights_string
+                if rights & 0b1000 != 0 {
+                    rights_string.push('K');
+                }
+                if rights & 0b0100 != 0 {
+                    rights_string.push('Q');
+                }
+                if rights & 0b0010 != 0 {
+                    rights_string.push('k');
+                }
+                if rights & 0b0001 != 0 {
+                    rights_string.push('q');
+                }
+
+                rights_string.push(' ');
+                rights_string
+            }
+        });
```

Comment:
instead of building a string for castling rights, you should set it equal to `rights_string = "KQkq"` and then `.filter()` to remove the unused ones. I think that would clean it up a lot.
If it comes up empty put in `-`